### DISCUSSION TITLE
Run code before hub dispatcher

### DISCUSSION THREAD

**Discussion by SylvainGantois** [dotnet/aspnetcore#58650] [Q&A]
We have a rate limiter that works for both HTTP requests and SignalR (we use exclusively web sockets for SignalR).

The SignalR rate limiting currently works as a hub filter, however, this will not take into account calls made to an inexisting hub method.
The default hub dispatcher displays a debug log and that's the end of the request:
![image](https://github.com/user-attachments/assets/b5445cbf-905e-471c-85cb-596cd4c9d107)

We have rate limits per user or SignalR connection and those calls should be taken into account (if a bad client makes a lot of invalid calls, we want to be aware of it, and maybe disable the user or block the IP, etc...).

Is there a way to create a custom hub dispatcher that would do some work before calling the default one, or a way to hook something else before the dispatcher?

Thank you.

**Comment by AshkanAfsharpour** [comment]
Have you tried overriding the default SignalR `HubDispatcher` with your custom implementation?

```csharp
services.AddSingleton(typeof(HubDispatcher<YourHubType>), typeof(CustomHubDispatcher<YourHubType>));
services.AddSingleton<IRateLimiter, YourRateLimiterImplementation>();

**Comment by SylvainGantois** [comment]
Unfortunately, it looks like we cannot use the HubDispatcher class anymore.
https://github.com/dotnet/aspnetcore/discussions/23508

Feels like a rabbit hole if we have to reimplement a bunch of classes.
